File: rx/followers/followers.cc
#include <iostream>
#include <string>

#include <rxcpp/rx.hpp>

using namespace std;

namespace rx = rxcpp;
namespace rxs = rxcpp::subjects;

using another_t = int;
using refresh_t = nullptr_t;
using startup_t = nullptr_t;

class Emitter
{
  rxs::subject<another_t> another_source;
  rxs::subject<refresh_t> refresh_source;
  rxs::subject<startup_t> startup_source;
public:
  rx::observable<another_t> another_stream;
  rx::observable<refresh_t> refresh_stream;
  rx::observable<startup_t> startup_stream;

  Emitter()
  {
    another_stream = another_source.get_observable();
    refresh_stream = refresh_source.get_observable();
    startup_stream = startup_source.get_observable();
  }

  void another(int n)
  {
    another_source.get_subscriber().on_next(n);
  }

  void refresh()
  {
    refresh_source.get_subscriber().on_next(nullptr);
  }

  void startup()
  {
    startup_source.get_subscriber().on_next(nullptr);
  }
};

auto connect(rx::observable<another_t> another_stream,
             rx::observable<refresh_t> refresh_stream,
             rx::observable<startup_t> startup_stream)
{
  return rx::observable<>::just(string("https://api.github.com/users"))
    .combine_latest(
      [](auto b, auto p) {
        return b + p;
      }
    ,
      startup_stream
        .map([](auto) {
          return string("");
        })
        .merge(
          refresh_stream
            .map([](auto) {
              return 1;
            })
            .scan(0, [](int a, int i) {
              return a + i;
            })
            .map([](int i) {
              return string("?since=") + to_string(i);
            })
        )
    );
}

int main()
{

  rxs::subject<int> s;

  s.get_observable().scan(1, [](int a, int i) {
    return a + i;
  });

  s.get_observable().subscribe([](int n) {
    cout << n << endl;
  });

  //s.get_subscriber().on_next(1);


  return 0;

  Emitter emit;

  connect(
    emit.another_stream,
    emit.refresh_stream,
    emit.startup_stream
  )
  .subscribe([](auto url) {

    cout << url << endl;
  });

  cout << "\033[32mstartup" << endl;
  emit.startup();

  cout << "\033[31mrefresh" << endl;
  emit.refresh();

  cout << "\033[33manother(1)" << endl;
  emit.another(1);

  cout << "\033[34manother(2)" << endl;
  emit.another(2);

  cout << "\033[35manother(0)" << endl;
  emit.another(1);

  cout << "\033[36manother(1)" << endl;
  emit.another(0);

  cout << "\033[37mrefresh" << endl;
  emit.refresh();

  return 0;
}

File: rx/followers/followers.py
import re

from functools import reduce
from pprint import pprint

import rx

import requests


RE_LINK = re.compile(r'<(.+)>; rel="next"')


class Emitter:

  def __init__(self):
    self.__anotherSource = rx.subjects.Subject()
    self.__refreshSource = rx.subjects.Subject()
    self.__startupSource = rx.subjects.Subject()

    self.anotherStream = self.__anotherSource.as_observable()
    self.refreshStream = self.__refreshSource.as_observable()
    self.startupStream = self.__startupSource.as_observable()

  def another(self, index):
    self.__anotherSource.on_next(index)

  def refresh(self):
    self.__refreshSource.on_next(None)

  def startup(self):
    self.__startupSource.on_next(None)


def main():
  emit = Emitter()
  followers_stream = connect(emit.anotherStream,
                             emit.refreshStream,
                             emit.startupStream)
  followers_stream.subscribe(pprint)
  print('\033[32mstartup')
  emit.startup()
  print('\033[33mrefresh')
  emit.refresh()
  print('\033[31manother 2')
  emit.another(2)
  print('\033[34manother 1')
  emit.another(1)
  print('\033[35manother 0')
  emit.another(0)
  print('\033[36manother 1')
  emit.another(1)
  print('\033[37mrefresh')
  emit.refresh()

  emit.refresh()
  emit.refresh()
  emit.refresh()
  emit.refresh()
  emit.refresh()
  emit.refresh()

  emit.refresh()
  emit.refresh()
  emit.refresh()


def connect(another_stream, refresh_stream, startup_stream):
  both_stream = startup_stream.merge(refresh_stream)
  fire_stream = (
    both_stream
      .flat_map(lambda _: rx.Observable.range(0, 3))
      .merge(another_stream)
  )

  return (
    rx.Observable
      .zip(
        fire_stream
      ,
        fire_stream
          .buffer_with_count(30)
          .start_with(None)
          .scan(
            lambda a, l:
              (
                lambda r: (
                  RE_LINK.match(r.headers['Link']).group(1),
                  r.json()
                )
              )(requests.get(a[0]))
            ,
            ('https://api.github.com/users', None)
          )
          .flat_map(lambda t: t[1])
          .map(lambda u: u['login'])
      ,
        lambda i, u: (i, u)
      )
      .buffer(both_stream.merge(another_stream))
      .scan(
        lambda a, e:
          reduce(lambda a, e: a[:e[0]] + [e[1]] + a[e[0]+1:], e, a)
        ,
        3 * [None]
      )
  )


if '__main__' == __name__:
  main()
